Implement ranges::copy_backward

CaseyCarter opened the pull request:
With many thanks to @ahanamuk who did much of the work.

Partially addresses #39.

miscco commented on tests/std/tests/P0896R4_ranges_alg_copy_backward/test.cpp:
```diff
@@ -0,0 +1,84 @@
+            { // Validate range overload
+                int output[] = {-2, -2, -2};
+                Bidi2 wrapped_output{output};
+                auto result = copy_backward(wrapped_input, wrapped_output.end());
```
We do not really test the one real reason we have for `copy_backward` namely overlapping regions. Should we test them? Especially for the memmove case

CaseyCarter replied:
Both `copy` and `copy_backward` handle overlapping regions, it's the direction of the allowed overlap that differs. Given:
```
|-----|-----|-----|
a     b     c     d
```
both `copy(b, d, a)` and `copy_backward(a, c, d)` are valid. 

The reason I don't have test coverage for overlapping ranges for `copy_backward` is the same reason we don't have test coverage for overlapping ranges for `copy`: both range algorithms are straightforward "rangifications" of our existing reliable `std` algorithm. In the case of `ranges::copy_backward` its literally implemented by the same lines of code. In cases where we reuse the `std` algorithm like this we've been performing fairly light coverage to ensure the adaptation is correct and reserving exhaustive testing of the underlying algorithm for cases where the range algorithm materially differs.

CaseyCarter replied:
Clarifying: this is my explanation for why there is no such test since you asked "Should we test them?" It's not a refusal to add such a test if you disagree with my rationale or simply think we should test it anyway.

StephanTLavavej replied:
I'm ok with this light testing as long as the core algorithm is literally centralized. (e.g. for `reverse` I was nervous about not seeing empty/odd/variously-vectorized cases, because the code wasn't totally centralized)

I'd also be fine with more testing; almost seems easier to write than to talk about.

miscco replied:
Yeah I guessed it would be handled like that.  Sound absolutely reasonable to only check new implementations.

I guess when the TMP machinery is overhauled to allow `contiguous_iterator`s rather than plain pointers this would be the place to add tests

CaseyCarter replied:
Overlapping test added.

miscco commented on tests/std/tests/P0896R4_ranges_alg_copy_backward/test.cpp:
```diff
@@ -0,0 +1,84 @@
+                STATIC_ASSERT(same_as<decltype(result), copy_backward_result<iterator_t<Bidi1>, iterator_t<Bidi2>>>);
+                assert(result.in == wrapped_input.end());
+                assert(result.out == wrapped_output.begin());
+                assert(ranges::equal(output, input));
```
Below you are pulling in ranges::equal. Same for the `call()`?

CaseyCarter replied:
Yes! No idea how I missed this.

miscco commented on tests/std/include/range_algorithm_support.hpp:
```diff
@@ -122,6 +122,26 @@ namespace test {
+        friend constexpr boolish operator==(sentinel const s, Element* const ptr) noexcept {
+            return {s.ptr_ == ptr};
+        }
+        friend constexpr boolish operator==(Element* const ptr, sentinel const s) noexcept {
+            return {s.ptr_ == ptr};
+        }
+        friend constexpr boolish operator!=(sentinel const s, Element* const ptr) noexcept {
+            return !(s == ptr);
+        }
+        friend constexpr boolish operator!=(Element* const ptr, sentinel const s) noexcept {
+            return !(s == ptr);
+        }
```
I believe (hope) those 3 are superfluous due to C++20 operator rewriting?

miscco replied:
Also I am confused. Do we use `east const` or `const west`?

CaseyCarter replied:
`test::iterator` and `test::sentinel` have to implement their own comparisons because they are ill-behaved and return `boolish`. C++20 comparison expression rewrites are only activated when the target of the rewrite has type `bool`. (I should say "intentionally ill-behaved" here since the point is to ensure that the algorithms work with ill-behaved user types.)

CaseyCarter replied:
> Also I am confused. Do we use `east const` or `const west`?

Internal consistency within a file is paramount, and this header was one of the last things I wrote before deciding to acquiesce to the prevalent `const`-west style in the test suite. I will eventually change this to `const`-west when we reach a point where I can make a global change like that in this header without conflicting with any in-flight PRs.

miscco replied:
Thanks for the explanation. 

I am (fortunately?) not yet accustomed with all those hairy corner cases.

miscco commented on tests/std/tests/P0896R4_ranges_alg_copy_backward/test.cpp:
```diff
@@ -0,0 +1,84 @@
+                auto result = copy_backward(wrapped_input, wrapped_output.end());
+                STATIC_ASSERT(same_as<decltype(result), copy_backward_result<iterator_t<Bidi1>, iterator_t<Bidi2>>>);
+                assert(result.in == wrapped_input.end());
+                assert(result.out == wrapped_output.begin());
```
Dont we need `peek()` here as `result.out` is not a sentinel?

CaseyCarter replied:
We only need to break the abstraction with `peek` when we have an input or output iterator and no sentinel to compare it with. The most common examples being a lone output iterator or an input iterator that an algorithm advances only partway through its range (so we have _a_ sentinel, but not one in the correct position.)

miscco commented on stl/inc/algorithm:
```diff
@@ -1567,6 +1567,39 @@ namespace ranges {
+            const auto _UFirst = _Get_unwrapped(_STD move(_First));
+            auto _ULast        = _Get_final_iterator_unwrapped<_It1>(_UFirst, _STD move(_Last));
+            _Seek_wrapped(_First, _ULast);
+            _Result = _Copy_backward_unchecked(_STD move(_UFirst), _STD move(_ULast), _STD move(_Result));
```
Cant we use `_Get_unwrapped_unverified(_STD move(_Result))` here?

CaseyCarter replied:
This will eventually use `_Get_unwrapped_n(_Result, _Idl_distance(_UFirst, _ULast))`. While we *could* use `_Get_unwrapped_unverified` without harming correctness I prefer to avoid unwrapping completely to make the things that will need fixing later easier to spot.

CaseyCarter replied:
I just rewrote the issue description in #893 with some details of the iterator unwrapping machinery that hopefully clarify my response.

miscco replied:
Thanks a lot :+1:

CaseyCarter commented on tests/std/include/range_algorithm_support.hpp:
```diff
@@ -445,11 +466,19 @@ namespace test {
 
         constexpr void _Seek_to(unwrap&& i) noexcept requires (to_bool(Wrapped)) {
-            ptr_ = i.peek();
+            if constexpr (at_least<contiguous>) {
```
Not that this line of code will see, but in principle users can define their own iterator categories. It's better to consistently think about which std iterator tags an iterator category is derived from.

miscco replied:
Thanks for the explanation. I fuly agree

CaseyCarter commented on tests/std/tests/P0896R4_ranges_alg_copy_backward/test.cpp:
```diff
@@ -0,0 +1,84 @@
+
+    { // Validate range overload
+        int output[] = {-2, -2, -2};
+        auto result  = copy_backward(instantiator::input, end(output));
```
I'm a huge proponent of DRY, but this usage is pretty awkward. I'll promote `instantiator::input` to the global namespace.

StephanTLavavej commented on tests/std/include/range_algorithm_support.hpp:
```diff
@@ -432,7 +452,8 @@ namespace test {
         using _Prevent_inheriting_unwrap = iterator;
 
-        using unwrap = iterator<Category, Element, Diff, Eq, Proxy, IsWrapped::no>;
+        using unwrap = std::conditional_t<derived_from<Category, contiguous>, Element*,
+            iterator<Category, Element, Diff, Eq, Proxy, IsWrapped::no>>;
```
This appears to be causing test failures:
```
C:\agent\_work\1\s\tests\std\tests\P0896R4_ranges_test_machinery\test.cpp(47): error C2338: same_as<_Unwrapped_t<I>, iterator<Category, Element, Diff, Eq, Proxy, IsWrapped::no>>
C:\agent\_work\1\s\tests\std\tests\P0896R4_ranges_test_machinery\test.cpp(139): note: see reference to function template instantiation 'bool iter_test<std::contiguous_iterator_tag,int,test::CanDifference::no,test::CanCompare::yes,test::ProxyRef::no,test::IsWrapped::yes>(void)' being compiled
```

CaseyCarter replied:
Yes. That's easily fixed - but the ICEs in the heap algorithm test are more challenging. It's the same "`constexpr` reference-to-pointers" bug that plagues many of the permuting algorithms, I'm narrowing down exactly which bits to snip out.

StephanTLavavej replied:
I've added a comment to Microsoft-internal VSO-938163 asking for this bug to be considered equal priority with C++20 Core Language feature work, since `std::ranges` is C++20.

CaseyCarter replied:
Workaround (actually avoidance) applied. The FE team has told me that they are working on VSO-938163 now in hopes of getting it fixed for 16.8. I'll move this PR to Initial Review and un-Draft it once the tests pass.

StephanTLavavej commented:
I'll second @miscco's concern that this fuses together changes which could apparently be separate. If they could be disentangled, it would be easier to review (and serve as a good example for contributors). If that's more work than it's worth, then it's okay in the name of expediency.

StephanTLavavej commented on stl/inc/algorithm:
```diff
@@ -1567,6 +1567,39 @@ namespace ranges {
+            requires indirectly_copyable<_It1, _It2>
+        constexpr copy_backward_result<_It1, _It2> operator()(_It1 _First, _Se1 _Last, _It2 _Result) const {
+            _Adl_verify_range(_First, _Last);
+            const auto _UFirst = _Get_unwrapped(_STD move(_First));
```
`_UFirst` is `const`, but it's followed by `_STD move(_UFirst)`.

StephanTLavavej commented on stl/inc/yvals_core.h:
```diff
@@ -346,7 +346,7 @@
 #endif
 
-// Controls whether the STL uses "if constexpr" internally
+// Controls whether the STL uses "if constexpr" internally in sub-C++17 modes
```
There's only one:
```suggestion
// Controls whether the STL uses "if constexpr" internally in C++14 mode
```

CaseyCarter replied:
We have enough compiler flags that there are trillions of C++14 modes! (Change accepted.)

StephanTLavavej commented on tests/std/include/range_algorithm_support.hpp:
```diff
@@ -122,6 +122,26 @@ namespace test {
         }
         // clang-format on
+
+        friend constexpr boolish operator==(sentinel const s, Element* const ptr) noexcept {
```
Should these operators be `[[nodiscard]]`?

CaseyCarter replied:
Yes. There a ton of operators in here that should be `[[nodiscard]]`.

StephanTLavavej commented on tests/std/tests/P0896R4_ranges_alg_copy_backward/test.cpp:
```diff
@@ -0,0 +1,84 @@
+
+constexpr void test_memmove() {
+    // Get some coverage for the memmove optimization, which we would not otherwise have since we do not (as of
+    // 20200709) unwrap output iterators.
```
I'd recommend 2020-07-09 as easier to read. I'd also recommend directly citing GH-893, especially because the GitHub Issues and Pull Requests extension for VSCode will recognize such citations in its next update.

CaseyCarter commented:
> What I do not like ( and I mean like in the sense that I do not have any technical objection) is that it conflates 3 different things.
> 
> 1. It implements the algortihms  `copy_backward`
> 2. It simplifies some test machinery
> 3. It introduces the TMP optimizations for ranges algorithms.
> 
> That not only makes this PR way more complex that the actual algorithm but also introduces a lot of inconsistencies that would have been better addressed with separate PRs.

This is a righteous comment: I let this one get away from me. I'll try to submit more quantized PRs in the future. (And by "in the future" I mean "after `reverse` ;))

StephanTLavavej commented on tests/std/include/range_algorithm_support.hpp:
```diff
@@ -203,70 +203,74 @@ namespace test {
 
         // clang-format off
-        friend constexpr boolish operator==(proxy_reference ref, Value const& val) requires CanEq<Element, Value> {
+        [[nodiscard]] friend constexpr boolish operator==(proxy_reference ref, Value const& val) requires CanEq<Element, Value> {
```
These lines are exceeding 120 columns :scream_cat:

CaseyCarter replied:
Wot? How??!? Cursed VS Code.

CaseyCarter replied:
Oh, derp - `// clang-format off`.

CaseyCarter replied:
Fixed.

StephanTLavavej commented:
It appears that this is triggering an ICE with our Microsoft-internal "chk" compiler (i.e. internal assertions enabled):

```
P0896R4_ranges_alg_move\test.cpp(73): Assertion failed: value->is_object(),
file S:\msvc\src\vctools\Compiler\CxxFE\sl\p1\c\constexpr\constexpr.cpp, line 6496
```

I'm going to exclude this PR from the current batch; we'll need to investigate this later.

StephanTLavavej commented:
I pushed a merge to resolve a simple conflict between the `[[nodiscard]]` addition and a rename from `ref` to `r`.

CaseyCarter commented:
**** NOTE TO REVIEWERS ***

I squashed-and-rebased what was here onto master to semi-reset, and added a new commit with additional changes. You should be ably to quickly scan the first commit to verify that it contains everything up though Stephan's merge conflict fix, and then focus on the changes in the second commit. That second commit primarily removes all earlier attempts to avoid triggering GH-1030 and applies a new localized workaround to avoid testing `constexpr` when `/analyze` is enabled.

StephanTLavavej commented on tests/std/tests/P0896R4_ranges_alg_move/test.cpp:
```diff
@@ -36,40 +36,45 @@ struct instantiator {
     static constexpr void call() {
-        using ranges::move, ranges::move_result, ranges::iterator_t;
+#if !defined(__clang__) && !defined(__EDG__) // TRANSITION, VSO-938163
+#pragma warning(suppress : 4127) //  conditional expression is constant
```
```suggestion
#pragma warning(suppress : 4127) // conditional expression is constant
```

StephanTLavavej commented on tests/std/tests/P0896R4_ranges_test_machinery/test.cpp:
```diff
@@ -26,9 +26,8 @@ constexpr bool iter_test() {
-    constexpr bool can_write =
-        derived_from<Category,
-            output_iterator_tag> || (derived_from<Category, forward_iterator_tag> && assignable_from<Element&, Element>);
+    constexpr bool can_write = derived_from<Category, output_iterator_tag> //
```
This is intentional (and mysterious-looking) - it forces clang-format to wrap the line. There are a few occurrences of this where clang-format can do something reasonable with a bit of assistance, without having to completely suppress it.

mnatsuhara replied:
I see...I was wondering if it was something like that!

CaseyCarter replied:
Yep: this is me forcing a reasonable line-break before the binary operator instead of allowing clang-format to break it in the middle of a perfectly good template-id. (Related: My belief that `requires` followed by `(` should be formatted differently in:
```c++
template <class T>
    requires (meow<T> + bar<T> < 42)
void f(T) { /* ... */ }
```
vs.
```c++
template <class T>
concept C =  requires(T x) { x + 42; );
```
has apparently delayed the clang-format concepts-and-requires-expressions patch enough to make it miss the Clang 11 fork. :sadpanda:)

CaseyCarter commented:
Thanks for you contribution!
!noitubirtnoc uoy rof sknahT
